File: benchmarks/pseudo-perfect/lib/pseudo-perfect.pl
% Create pseudo-perfect numbers (numbers of the form 2^(p-1)*(2^p - 1), where p is prime.
% Based on perfect.pl from the van Roy benchmark set (https://github.com/SWI-Prolog/bench)

top :-
	findall(C, perfect(C), X),
    print(X),
	ok(X).

ok([ 6,
     28,
     496,
     8128,
     2096128,
     33550336,
     8589869056,
     137438691328,
     35184367894528,
     144115187807420416,
     2305843008139952128,
     9444732965670570950656,
     2417851639228158837784576,
     38685626227663735544086528,
     9903520314282971830448816128,
     40564819207303336344294875201536,
     166153499473114483824745506383331328,
     2658455991569831744654692615953842176,
     10889035741470030830754200461521744560128,
     2787593149816327892690784192460327776944128,
     44601490397061246283066714178813853366747136,
     182687704666362864775460301858080473799697891328,
     46768052394588893382517909811217778170473142550528,
     191561942608236107294793378084303638130997321548169216,
     12554203470773361527671578846336104669690446551334525075456,
     3213876088517980551083924184681057554444177758164088967397376
   ]).

isComposite(X, Y) :-
	N is Y*Y,
	N =< X,
	X mod Y =:= 0.
isComposite(X, Y) :-
	Y < X,
	Y1 is Y + 1,
	isComposite(X, Y1).

% Non-deterministically bind X to every prime number in the given list
selectPrime([X|_], X) :-
	Y is 2,
	X > 1,
	\+ isComposite(X, Y).
selectPrime([_|T], Z) :-
	selectPrime(T, Z).

power(_, 0, 1) :- !.
power(B, P, R) :-
	P1 is P-1,
	power(B, P1, R1),
	R is R1*B.

% compute 2^(p-1)*(2^p-1)
pseudoPerfect(P, R) :-
	power(2, P, X),
	R1 is X-1,
	power(2, P-1, R2),
	R is R1 * R2.

% Non-deterministically bind R to the pseudo-perfect number generated by each prime in the given list
selectPerfect([P|_], R) :-
	pseudoPerfect(P, R).
selectPerfect([_|T], R) :-
	selectPerfect(T, R).


%generate one list of N numbers.
%genList(10, L).
generateList(N, Xs) :- generateList(0, N, Xs).
generateList(N, N, [N]).
generateList(X, N, [X|Xs]) :-
    X < N,
    X1 is X + 1,
    generateList(X1, N, Xs).

%list of N perfect numbers
%perfect(100, C)
perfect(C) :-
	generateList(101, R),
	findall(L, selectPrime(R, L), P),
	selectPerfect(P, C).

File: benchmarks/pseudo-perfect/lib/PseudoPerfect.hs
module PseudoPerfect where

import Control.Hspl
import Data.Numbers.Primes

maxCandidatePrime :: Integer
maxCandidatePrime = 101

isComposite :: Predicate Integer
isComposite = predicate "isComposite" $
  match(integer "x") |- go? (v"x", 2 :: Integer)
  where go :: Predicate (Integer, Integer)
        go = predicate "isComposite.go" $ do
          match(v"x", v"y") |- do
            integer "x" |>=| v"y" |*| v"y"
            v"x" |%| v"y" |==| (0 :: Integer)
          match(v"x", v"y") |- do
            integer "y" |<| v"x"
            go? (v"x", v"y" |+| (1 :: Integer))

selectPrime :: Predicate ([Integer], Integer)
selectPrime = predicate "selectPrime" $ do
  match(v"x" <:> __, v"x") |- do
    v"x" |>| (1 :: Integer)
    lnot $ isComposite? v"x"
  match(__ <:> v"xs", v"y") |-
    selectPrime? (v"xs", v"y")

power :: Predicate (Integer, Integer, Integer)
power = predicate' [SemiDet] "power" $ do
  match(__, 0 :: Integer, 1 :: Integer)
  match(v"b", v"p", v"r") |- do
    v"p1" |==| v"p" |-| (1 :: Integer)
    power? (v"b", v"p1", v"r1")
    integer "r" |==| v"r1" |*| v"b"

-- | Compute 2^(p-1)(2^p - 1)
pseudoPerfect :: Predicate (Integer, Integer)
pseudoPerfect = predicate "pseudoPerfect" $
  match(v"p", v"r") |- do
    power?(2 :: Integer, v"p", v"2^p")
    v"r1" |==| v"2^p" |-| (1 :: Integer)
    power?(2 :: Integer, v"p" |-| (1 :: Integer), v"r2")
    integer "r" |==| v"r1" |*| v"r2"

selectPerfect :: Predicate ([Integer], Integer)
selectPerfect = predicate "selectPerfect" $ do
  match(v"p" <:> __, v"r") |- pseudoPerfect? (v"p", v"r")
  match(__ <:> v"ps", v"r") |- selectPerfect? (v"ps", v"r")

perfect :: Predicate Integer
perfect = predicate "perfect" $
  match(v"c") |- do
    findAll (integer "p") (selectPrime?([0..maxCandidatePrime], v"p")) (v"ps")
    selectPerfect?(v"ps", v"c")

expected :: [Integer]
expected = [2^(p-1) * (2^p - 1) | p <- [0..maxCandidatePrime], isPrime p]
